### PR TITLE
[RF] MultiProcess error handling

### PR DESCRIPTION
**Pull request by egpbos**
# This Pull request:

## Changes or fixes:
This PR mainly adds error handling to the `MultiProcess`-based fitting stack. Error signaling in RooFit happens through two mechanisms: `logEvalError` calls and `RooNaNPacker`-enhanced `nan` doubles. Both are now implemented and working.

With this addition, we can now (finally!) run the most recent ATLAS Higgs combination fits again without issues.

The actual error handling is implemented in the 5th commit of this PR.

The first two commits are really unrelated, minor refactorings of existing code. I add them in this PR because I touched these while working on the error handling and just wanted to rebase to keep history clean and don't think they need a separate PR, since they are quite trivial changes.

The same goes for the last commit; it's a refactoring that could have been done in a different PR, but I hope you will permit this shortcut for these three trivial commits ;)

Finally, commit 3 and 4 add tests. I kept these commits separate mainly for reviewing and also for my testing convenience (I wanted to double check they indeed fail before applying commit 5). If you prefer, we can squash commits 3, 4 and 5 before we merge.

## Checklist:

- [x] tested changes locally
- [ ] ~updated the docs (if necessary)~ _There are no user-facing changes (except that more things now work), everything happens behind the scenes, so no documentation updates are necessary._

Thanks to @Zeff020 and @guitargeek for help in getting this done!

### CODE REVIEW

File: roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx [hunk 1]
```diff
@@ -22,15 +22,26 @@
...
 #include "RooFit/TestStatistics/LikelihoodWrapper.h"
 #include "RooFit/TestStatistics/RooUnbinnedL.h"
 #include "RooFit/TestStatistics/RooRealL.h"
 #include "RooFit/TestStatistics/buildLikelihood.h"
 #include "RooFit/MultiProcess/JobManager.h"
 #include "RooFit/MultiProcess/Config.h"
 
+// for MinuitFcnGrad test:
+#include "RooFit/MultiProcess/Config.h"
+#include "RooStats/ModelConfig.h"
```
Comment (guitargeek) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx:35]:
```suggestion
#include "RooFit/ModelConfig.h"
```

The ModelConfig was moved from RooFit to RooStats, because it is used so much also outside the context of RooStats. This should fix the CI build failures

Comment (egpbos) [reply roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx]: Awesome, thanks, fixed and force-pushed.

**Comment by guitargeek** [comment]
Thank you very much for your work, @egpbos! That's amazing progress!

Do you know what's going on with these failures on macOS?

**Comment by egpbos** [comment]
Unfortunately not. The fact that it's ARM and that we see half of those same failures also on the march=native alma9 build makes me suspect something to do with a wrong assumption on floating point arithmetic. However, the issue for those march=native failures (and the same failing 6 tests on mac ARM, which all seem to be small floating point errors) may be a different one than the one causing the other 6 failing tests on mac which are more mysterious (huge difference there).

**Comment by egpbos** [comment]
I just added an additional commit that adds functionality around error handling. It catches a corner case that may be relevant to the failing tests (so I hope it will add useful information from the prints in the CI runs), but that I think in any case should be checked for, because it's about a fit that failed but that Minuit reports as successful. See commit message for more info.

**Comment by egpbos** [comment]
While fixing the CI runs, I reordered the commits a bit as well:

1. The first two commits are still separate from the main part of this PR, see description in first message.
3. The third commit is the actual meat: it adds error handling to MultiProcess, including tests.
4. The fourth commit is a bug fix, but since it is a rather rare bug that is only triggered in pathological cases (that I happened to trigger in one of the tests added in commit 3) and since it interacts with a specific other part (cache-and-tracking), I thought it was better to keep it separate, so that the commit can explain the rationale for that specific fix well.
5. Commit 5 is again a minor refactoring to reduce some duplicated code.
6. Commit 6 adds a strong warning message when fits end after only having tried parameters that give evaluation errors. In such a case, there will not actually be any valid value, but rather the error handling code will return zero as the function value. Minuit sees this as a valid function value and thinks the function converged and so RooFit would also think everything ended well. With this commit, at least a warning is issued. Perhaps this response should be stronger or weaker, that is up for discussion.
7. Commit 7 applies a recommendation from the Form documentation to replace it with Format.
8. Code formatting.

**Comment by egpbos** [comment]
`alma9 march_native` CI failure not related to this PR (`roottest-root-hist-h2root` fails).

**Comment by egpbos** [comment]
Ready for review :)

File: roofit/roofitcore/inc/RooMinimizer.h [hunk 2]
```diff
@@ -39,6 +39,13 @@ class RooRealVar;
 class RooArgSet;
 class RooPlot;
 class RooDataSet;
+#ifdef ROOFIT_MULTIPROCESS
+namespace RooFit {
+namespace TestStatistics {
+class LikelihoodGradientJob;
+}
+} // namespace RooFit
+#endif // ROOFIT_MULTIPROCESS
```
Comment (guitargeek) [inline roofit/roofitcore/inc/RooMinimizer.h:48]:
```suggestion
namespace RooFit {
namespace TestStatistics {
class LikelihoodGradientJob;
}
} // namespace RooFit
```

`ROOFIT_MULTIPROCESS` is only defined when building ROOT, so the macros like this should not be used in public header files. Here, since you are only introducing a forward declaration, can we just always do it?

File: roofit/roofitcore/inc/RooMinimizer.h [hunk 3]
```diff
@@ -162,6 +163,9 @@ class RooMinimizer : public TObject {
 private:
    friend class RooAbsMinimizerFcn;
    friend class RooMinimizerFcn;
+#ifdef ROOFIT_MULTIPROCESS
+   friend class RooFit::TestStatistics::LikelihoodGradientJob;
+#endif // ROOFIT_MULTIPROCESS
```
Comment (guitargeek) [inline roofit/roofitcore/inc/RooMinimizer.h:168]:
```suggestion
   friend class RooFit::TestStatistics::LikelihoodGradientJob;
```
Same as the other comment.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx [hunk 4]
```diff
@@ -76,10 +76,17 @@ RooAbsMinimizerFcn::RooAbsMinimizerFcn(RooArgList paramList, RooMinimizer *conte
...
-     _initFloatParamList(std::make_unique<RooArgList>()), _initConstParamList(std::make_unique<RooArgList>()),
+   : _context(other._context),
+     _maxFCN(other._maxFCN),
+     _funcOffset(other._funcOffset),
+     _numBadNLL(other._numBadNLL),
+     _evalCounter(other._evalCounter),
+     _nDim(other._nDim),
+     _optConst(other._optConst),
+     _floatParamList(new RooArgList(*other._floatParamList)),
+     _constParamList(new RooArgList(*other._constParamList)),
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:87]:
```suggestion
     _floatParamList(std::make_unique<RooArgList>(*other._floatParamList)),
     _constParamList(std::make_unique<RooArgList>(*other._constParamList)),
```
If you are already touching these lines :)

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx [hunk 5]
```diff
@@ -442,6 +449,36 @@ void RooAbsMinimizerFcn::printEvalErrors() const
...
 
+/// Apply corrections on the fvalue if errors were signaled.
+///
+/// Two kinds of errors are possible: 1. infinite or nan values (the latter
+/// can be a signaling nan, using RooNaNPacker) or 2. logEvalError-type errors.
+/// Both are caught here and fvalue is updated so that Minuit in turn is nudged
+/// to move the search outside of the problematic parameter space area.
+double RooAbsMinimizerFcn::applyEvalErrorHandling(double fvalue) const
+{
+   if (!std::isfinite(fvalue) || RooAbsReal::numEvalErrors() > 0 || fvalue > 1e30) {
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:460]: Can you explain in an inline comment where the magic number `1e30` comes from?

File: roofit/roofitcore/src/RooAbsMinimizerFcn.h [hunk 6]
```diff
@@ -118,6 +119,10 @@ class RooAbsMinimizerFcn {
    mutable double _funcOffset{0.};
    mutable int _numBadNLL = 0;
    mutable int _evalCounter{0};
+   // PB: these mutables signal a suboptimal design. A separate error handling
```
Comment (guitargeek) [inline roofit/roofitcore/src/RooAbsMinimizerFcn.h:122]: Nice observation! You're planning to do this in the future?

Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.h:122]: For me it currently looks like a nice-to-have. Factoring it out would make the RooAbsMinimizerFcn simpler and better maintainable. For the MinuitFcnGrad, it would enable easier synchronization of these error-handling related parameters between calculators (likelihood and gradient) and also easier synchronization between master and workers in the parallel calculators. The latter is currently in fact not being done at all, i.e. evalCounter does not count all the evaluations done on workers. However, I don't know how many people actually use the evalCounter or numBadNLL, so it's not clear whether this is a need-to-have.

So, no, not planned from my side, not high enough priority for me (according to what I know currently at least). I would love to help out someone who does have time for it.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx (hunk 5 above)
Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:460]: As elaborated via mail, I cannot. It is a historical choice that I just copied over. Perhaps Wouter can elaborate.

Comment (egpbos) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:460]: Note that this change happened in the single commit https://github.com/root-project/root/pull/15797/commits/ae8c120ceff8d114aff6ad028160d9b8b3978486 so it should be clear to future maintainers that this was indeed simply copied over. The origin of the number can be traced back through git history.

File: roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx [hunk 7]
```diff
@@ -589,3 +588,413 @@ TEST_P(LikelihoodGradientJobTest, Gaussian1DAlsoWithLikelihoodJob)
...
+
+   m1.minimize("Minuit2", "migrad");
+
+   std::unique_ptr<RooFitResult> m1result{m1.save()};
+   double minNll1 = m1result->minNll();
+   double mu_sig1 = w.var("mu_sig")->getVal();
+   double mu_bkg_A1 = w.var("mu_bkg_A")->getVal();
+   double mu_bkg_B1 = w.var("mu_bkg_B")->getVal();
+
+   EXPECT_EQ(minNll0, minNll1);
```
Comment (guitargeek) [inline roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx:899]:
```suggestion
   EXPECT_DOUBLE_EQ(minNll0, minNll1);
```

This comparison fails on `mac-beta`:
```txt
2024-06-26T15:15:19.6793790Z /Users/sftnight/ROOT-CI/src/roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx:899: Failure
2024-06-26T15:15:19.6794320Z Expected equality of these values:
2024-06-26T15:15:19.6794520Z   minNll0
2024-06-26T15:15:19.6794660Z     Which is: 110.212
2024-06-26T15:15:19.6794820Z   minNll1
2024-06-26T15:15:19.6794960Z     Which is: 110.212
```
Maybe changing to `EXPECT_DOUBLE_EQ` would fix it?

Comment (egpbos) [reply roofit/roofitcore/test/TestStatistics/testLikelihoodGradientJob.cxx:899]: Yes, thanks, I'm on it. I think the fix is actually to (again) switch to component-based splitting. (We should really change the default from event to component based splitting, but that's for a next PR.) I will push it later.

**Comment by egpbos** [comment]
Switching to component-based splitting indeed fixed it. Will rebase the fixup commit into commit three now and force push the rebased commits.

File: roofit/roofitcore/src/RooAbsMinimizerFcn.cxx (hunk 5 above)
Comment (guitargeek) [reply roofit/roofitcore/src/RooAbsMinimizerFcn.cxx:460]: Ah yes, that makes sense. I didn't realize at first that the code was only moved.

**Review by guitargeek — APPROVED** [review]
Thanks a lot for these updates and addressing the review comments!

Good to know that the recent ATLAS Higgs combinations fits are running now :+1: :+1:
